{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fb4911f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/glennrd/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.5.16\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scipy as sci\n",
    "import rpy2\n",
    "print(rpy2.__version__)\n",
    "import gseapy as gp\n",
    "from xlrd import XLRDError\n",
    "import xlsxwriter\n",
    "from gprofiler import gprofiler\n",
    "import rpy2.robjects as ro\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d90034b3",
   "metadata": {},
   "source": [
    "## Differential Gene Expression Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bdf395a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_crypts = sc.read_h5ad('/home/glennrd/Documents/Research_Project/RNA-seq_Analysis/python/crypt/crypt_enriched.h5ad')\n",
    "adata_villi = sc.read_h5ad('/home/glennrd/Documents/Research_Project/RNA-seq_Analysis/python/villi/villus_enriched.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2da8eb36",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create batch annotation for limma (Crypts)\n",
    "adata_crypts.obs['Batch'] = ['batch'+x.split('_')[1] for x in adata_crypts.obs['Sample']]\n",
    "adata_crypts.obs['Batch'] = adata_crypts.obs['Batch'].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9c5ae0d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create batch annotation for limma (villi)\n",
    "adata_villi.obs['Batch'] = adata_villi.obs['diet'].map({'CD': 'batch1', 'HFD': 'batch2'})\n",
    "adata_villi.obs['Batch'] = adata_villi.obs['Batch'].astype('category')\n",
    "adata_villi.obs.rename(columns={'diet': 'Diet'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b836fac3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def abs_log2fc(adata,grouptype,group2,group1):\n",
    "    # calculate the absolute log2-foldchange as log2(b/a) = (ln(b)-ln(a))/ln(2)\n",
    "    # data is ln(counts+1), neglect the +1\n",
    "    obs_df=pd.DataFrame(adata.X.toarray(),columns=adata.var_names)\n",
    "    obs_df[grouptype]=adata.obs[grouptype].values    \n",
    "    grouped_df=obs_df.groupby([grouptype])\n",
    "    tot=grouped_df[adata.var_names].apply(lambda x: np.mean(x, axis=0))\n",
    "    print(\"Group2:\", group2, \"in tot:\", group2 in tot.index)\n",
    "    print(\"Group1:\", group1, \"in tot:\", group1 in tot.index)\n",
    "    lfc= (tot.loc[group2] - tot.loc[group1])/np.log(2)\n",
    "    return lfc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "27ba6410",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_limma_counts(adata, condobs, batchobs, countsobs, coef,ref):\n",
    "\n",
    "    \"\"\"\n",
    "    This function performs differential gene expression analysis using the limma and edgeR packages from R.\n",
    "    It takes an AnnData object, column names for condition, batch, counts, a coefficient for the limma model, \n",
    "    and a reference level for the condition factor. It returns a DataFrame containing the results of the \n",
    "    differential expression analysis.\n",
    "    \"\"\"\n",
    "\n",
    "    import rpy2.robjects as robjects\n",
    "    import rpy2.robjects.numpy2ri\n",
    "    from rpy2.robjects import pandas2ri\n",
    "    from rpy2.robjects.packages import importr\n",
    "    pandas2ri.activate()\n",
    "    \n",
    "    # prepare data for R\n",
    "    data=pd.DataFrame(adata.X.toarray(), columns=adata.var_names)\n",
    "    cond=adata.obs[condobs]\n",
    "    batch=adata.obs[batchobs]\n",
    "    counts=adata.obs[countsobs]\n",
    "\n",
    "    # load R packages and data\n",
    "    R=robjects.r\n",
    "    print('Loading edgeR package')\n",
    "    R('library(edgeR)')\n",
    "    print('Loading limma package')\n",
    "    R('library(limma)')\n",
    "    print('Assigning data ⌛')\n",
    "    R.assign('data',data.T)\n",
    "    #print(dgelist)\n",
    "    print('Assigning condition')\n",
    "    R.assign('cond', cond)\n",
    "    print('Assigning batch')\n",
    "    R.assign('batch', batch)\n",
    "    print('Assigning counts')\n",
    "    R.assign('counts', counts)\n",
    "    R('counts<-scale(counts)')\n",
    "\n",
    "    # delete for memory\n",
    "    del data\n",
    "    del cond    \n",
    "    del batch\n",
    "    \n",
    "    # format data and create dge object \n",
    "    print('Formatting data')\n",
    "    R('cond <- as.factor(cond)')\n",
    "    print('Releveling condition')\n",
    "    R('cond <- relevel(cond,'''+ref+''')''')\n",
    "    print('Creating DGEList')\n",
    "    R('dge <- edgeR::DGEList(data,group=cond,sample=data.frame(batch=batch,counts=counts))')\n",
    "    print('Deleting data')\n",
    "    R('rm(data)')\n",
    "    print('Creating EList')\n",
    "    R('y <- new(\"EList\")')\n",
    "    print('Assigning DGEList')\n",
    "    R('y$E <- dge')\n",
    "    print('Deleting DGEList')\n",
    "    R('rm(dge)')\n",
    "    \n",
    "    # design matrix for testing\n",
    "    print('Creating design matrix')\n",
    "    R('design <- model.matrix(~cond+batch+counts)')\n",
    "    \n",
    "    # run limma\n",
    "    print('Running Limma lmFit ⌛')\n",
    "    R('fit <- limma::lmFit(y, design = design)')\n",
    "    R('rm(y)')\n",
    "    print('Running Limma eBayes ⌛')\n",
    "    R('fit <-  limma::eBayes(fit, trend = TRUE, robust = TRUE,)')\n",
    "    \n",
    "    # get results\n",
    "    print('Getting results')\n",
    "    strg='\"BH\"'\n",
    "    tt = R('''limma::topTable(fit,coef='''+coef+''',number = Inf,adjust.method='''+strg+''')''')\n",
    "    return tt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "052bd9ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_amb(adata):\n",
    "    # ambient genes for filtering, see processing notebook\n",
    "    ambient_genes=['Itln1','Spink4','Zg16','Lyz1','Defa21','Gm14851','Defa22','Gm15308','Gm15284',\n",
    "                   'Defa20','Gm15308','Gm14850','Gm7861','Defa17','AY761184', 'Ang4','Agr2','Clps','Tff3','Defa24','Fcgbp']\n",
    "    ix_amb_genes = np.in1d(adata.var_names,ambient_genes,invert=True)\n",
    "    return (ix_amb_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0619e460",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_genes(adata_filt, obs_name, group1, group2):\n",
    "    # Filter genes expressed in at least 10% of cells in either group\n",
    "    ix_group1 = np.isin(adata_filt.obs[obs_name], group1)\n",
    "    adata_sub_group1 = adata_filt[ix_group1].copy()\n",
    "    filter_1 = sc.pp.filter_genes(adata_sub_group1.X, min_cells=int(adata_sub_group1.n_obs * 0.1), inplace=False)[0]\n",
    "    del adata_sub_group1\n",
    "\n",
    "    ix_group2 = np.isin(adata_filt.obs[obs_name], group2)\n",
    "    adata_sub_group2 = adata_filt[ix_group2].copy()\n",
    "    filter_2 = sc.pp.filter_genes(adata_sub_group2.X, min_cells=int(adata_sub_group2.n_obs * 0.1), inplace=False)[0]\n",
    "    del adata_sub_group2\n",
    "    \n",
    "    ix_genes=[a or b for a, b in zip(filter_1,filter_2)]\n",
    "    \n",
    "    adata_filt = adata_filt[:,np.array(ix_genes)].copy()\n",
    "    \n",
    "    return adata_filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "da6f7700",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_to_csv(df, filename, group, prefix):\n",
    "    if len(group)>17:\n",
    "            group_short=group[0:17]\n",
    "            group_short=group_short.replace('/', '_')\n",
    "            group_short=group_short.replace(' ', '_')\n",
    "            \n",
    "            df.to_csv(filename + \"_\" + group_short + prefix + \".csv\")\n",
    "    else:\n",
    "            group=group.replace('/', '_')\n",
    "            group=group.replace(' ', '_')\n",
    "\n",
    "            df.to_csv(filename + \"_\" + group + prefix + \".csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "0c2c6602",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of unique cells in crypt data:\n",
      "Enterocyte progenitor\n",
      "Goblet cell\n",
      "EEC\n",
      "Goblet progenitor\n",
      "ISC\n",
      "not annotated\n",
      "Tuft progenitor\n",
      "Enterocyte\n",
      "Tuft cell\n",
      "EE progenitor\n",
      "Unkown progenitor\n",
      "Paneth cell\n",
      "Paneth progenitor\n",
      "\n",
      "\n",
      "List of unique cells in crypt data:\n",
      "Goblet cell\n",
      "Tuft cell\n",
      "Enterocyte progenitor\n",
      "Enterocyte\n",
      "EEC\n",
      "ISC\n"
     ]
    }
   ],
   "source": [
    "cells_crypts = adata_crypts.obs['major_cell_types'].unique()\n",
    "print('List of unique cells in crypt data:')\n",
    "for cell in cells_crypts:\n",
    "    print(cell)\n",
    "\n",
    "print('\\n')\n",
    "\n",
    "cells_villi = adata_villi.obs['major_cell_types'].unique()\n",
    "print('List of unique cells in villi data:')\n",
    "for cell in cells_villi:\n",
    "    print(cell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "77a98b0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to perform differential expression analysis\n",
    "def differential_expression_analysis(list_cells, adata, dataset, group):\n",
    "    # Change working directory to the appropriate path\n",
    "    os.chdir(f'/home/glennrd/Documents/Research_Project/RNA-seq_Analysis/python/{dataset}/differential_expression')\n",
    "\n",
    "    for cell in list_cells:\n",
    "        # Start differential testing for the cell type\n",
    "        print(f'Performing differential expression analysis of {cell} cells ⌛')\n",
    "\n",
    "        # Select the cell type\n",
    "        ix_cells = np.isin(adata.obs[group], [cell])\n",
    "\n",
    "        # Create a subset of the data containing only those cell types\n",
    "        adata_filt = adata[ix_cells, :].copy()\n",
    "        print(f'Subsetting {np.sum(ix_cells)} cells')\n",
    "\n",
    "        # Filter out ambient genes\n",
    "        print('Removing ambient genes ⌛')\n",
    "        ix_amb = filter_amb(adata) \n",
    "        adata_filt = adata_filt[:, ix_amb].copy()\n",
    "\n",
    "        # Filter out genes that are expressed in less than 10% of cells in either group (CD or HFD)\n",
    "        print('Removing genes expressed in less than 10% of cells ⌛')\n",
    "        adata_filt = filter_genes(adata_filt, 'Diet', 'CD', 'HFD')\n",
    "\n",
    "        # Run differential expression analysis using limma\n",
    "        print('Running Limma ⌛')\n",
    "\n",
    "        try:\n",
    "            # Run limma counts with 'diet' as the condition, 'batch' as the batch, 'n_genes' as the counts, \n",
    "            # 'condHFD' as the coefficient for the limma model, and 'CD' as the reference level for the condition factor\n",
    "            x = run_limma_counts(adata_filt, 'Diet', 'Batch', 'n_genes', '\"condHFD\"', '\"CD\"')\n",
    "\n",
    "        except rpy2.rinterface_lib.embedded.RRuntimeError:\n",
    "            # If no significant genes are found, print a message and continue\n",
    "            print('No significant genes found..')\n",
    "            continue\n",
    "\n",
    "        print('Data stored in variable x')\n",
    "\n",
    "        print('Calculating absolute lfc')\n",
    "\n",
    "        base_filename = 'diff_exp_CD_vs_HFD'\n",
    "\n",
    "        # Compute absolute log2 fold change\n",
    "        lfc = abs_log2fc(adata_filt, 'Diet', 'HFD', 'CD')\n",
    "        x = ro.conversion.rpy2py(x)\n",
    "        x = x.loc[lfc.index]\n",
    "        x['abs.log2FC'] = lfc.values\n",
    "        del lfc\n",
    "\n",
    "        # Write to csv\n",
    "        print(f'Writing to {base_filename}_{cell} ⌛')\n",
    "        write_to_csv(pd.DataFrame(x), base_filename, cell, '')\n",
    "        print('Done ✅')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "1664a735",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: limma\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Performing differential expression analysis of Goblet cell cells ⌛\n",
      "Subsetting 1001 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Coefficients not estimable: batchbatch2 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: Partial NA coefficients for 7211 probe(s) \n",
      "\n",
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning messages:\n",
      "\n",
      "R[write to console]: 1: \n",
      "R[write to console]: In out$var.prior[is.na(out$var.prior)] <- 1/out$s2.prior :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  number of items to replace is not a multiple of replacement length\n",
      "\n",
      "R[write to console]: 2: \n",
      "R[write to console]: In .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  Estimation of var.prior failed - set to default value\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_Goblet cell ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of Tuft cell cells ⌛\n",
      "Subsetting 1267 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Coefficients not estimable: batchbatch2 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: Partial NA coefficients for 8352 probe(s) \n",
      "\n",
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning messages:\n",
      "\n",
      "R[write to console]: 1: \n",
      "R[write to console]: In out$var.prior[is.na(out$var.prior)] <- 1/out$s2.prior :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  number of items to replace is not a multiple of replacement length\n",
      "\n",
      "R[write to console]: 2: \n",
      "R[write to console]: In .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  Estimation of var.prior failed - set to default value\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_Tuft cell ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of Enterocyte progenitor cells ⌛\n",
      "Subsetting 218 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Coefficients not estimable: batchbatch2 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: Partial NA coefficients for 9778 probe(s) \n",
      "\n",
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning messages:\n",
      "\n",
      "R[write to console]: 1: \n",
      "R[write to console]: In out$var.prior[is.na(out$var.prior)] <- 1/out$s2.prior :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  number of items to replace is not a multiple of replacement length\n",
      "\n",
      "R[write to console]: 2: \n",
      "R[write to console]: In .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  Estimation of var.prior failed - set to default value\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_Enterocyte progenitor ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of Enterocyte cells ⌛\n",
      "Subsetting 1413 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Coefficients not estimable: batchbatch2 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: Partial NA coefficients for 7901 probe(s) \n",
      "\n",
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning messages:\n",
      "\n",
      "R[write to console]: 1: \n",
      "R[write to console]: In out$var.prior[is.na(out$var.prior)] <- 1/out$s2.prior :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  number of items to replace is not a multiple of replacement length\n",
      "\n",
      "R[write to console]: 2: \n",
      "R[write to console]: In .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  Estimation of var.prior failed - set to default value\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_Enterocyte ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of EEC cells ⌛\n",
      "Subsetting 331 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Coefficients not estimable: batchbatch2 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: Partial NA coefficients for 8828 probe(s) \n",
      "\n",
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning messages:\n",
      "\n",
      "R[write to console]: 1: \n",
      "R[write to console]: In out$var.prior[is.na(out$var.prior)] <- 1/out$s2.prior :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  number of items to replace is not a multiple of replacement length\n",
      "\n",
      "R[write to console]: 2: \n",
      "R[write to console]: In .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  Estimation of var.prior failed - set to default value\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_EEC ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of ISC cells ⌛\n",
      "Subsetting 115 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Coefficients not estimable: batchbatch2 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: Partial NA coefficients for 11416 probe(s) \n",
      "\n",
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning messages:\n",
      "\n",
      "R[write to console]: 1: \n",
      "R[write to console]: In out$var.prior[is.na(out$var.prior)] <- 1/out$s2.prior :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  number of items to replace is not a multiple of replacement length\n",
      "\n",
      "R[write to console]: 2: \n",
      "R[write to console]: In .ebayes(fit = fit, proportion = proportion, stdev.coef.lim = stdev.coef.lim,  :\n",
      "R[write to console]: \n",
      " \n",
      "R[write to console]:  Estimation of var.prior failed - set to default value\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_ISC ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of Enterocyte progenitor cells ⌛\n",
      "Subsetting 8325 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_Enterocyte progenitor ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of Goblet cell cells ⌛\n",
      "Subsetting 4539 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_Goblet cell ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of EEC cells ⌛\n",
      "Subsetting 2197 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_EEC ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of Goblet progenitor cells ⌛\n",
      "Subsetting 1844 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_Goblet progenitor ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of ISC cells ⌛\n",
      "Subsetting 3943 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_ISC ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of not annotated cells ⌛\n",
      "Subsetting 712 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_not annotated ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of Tuft progenitor cells ⌛\n",
      "Subsetting 476 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_Tuft progenitor ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of Enterocyte cells ⌛\n",
      "Subsetting 2598 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_Enterocyte ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of Tuft cell cells ⌛\n",
      "Subsetting 816 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_Tuft cell ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of EE progenitor cells ⌛\n",
      "Subsetting 254 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_EE progenitor ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of Unkown progenitor cells ⌛\n",
      "Subsetting 375 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_Unkown progenitor ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of Paneth cell cells ⌛\n",
      "Subsetting 1258 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_Paneth cell ⌛\n",
      "Done ✅\n",
      "Performing differential expression analysis of Paneth progenitor cells ⌛\n",
      "Subsetting 82 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Assigning counts\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Running Limma eBayes ⌛\n",
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_Paneth progenitor ⌛\n",
      "Done ✅\n"
     ]
    }
   ],
   "source": [
    "# Define variables for adata_villi\n",
    "list_cells_villi = cells_villi\n",
    "dataset_villi = 'villi'\n",
    "group_villi = 'major_cell_types'\n",
    "\n",
    "# Define variables for adata_crypts\n",
    "list_cells_crypts = cells_crypts\n",
    "dataset_crypts = 'crypt'\n",
    "group_crypts = 'major_cell_types'\n",
    "\n",
    "# Run differential expression analysis for adata_villi\n",
    "differential_expression_analysis(list_cells_villi, adata_villi, dataset_villi, group_villi)\n",
    "\n",
    "# Run differential expression analysis for adata_crypts\n",
    "differential_expression_analysis(list_cells_crypts, adata_crypts, dataset_crypts, group_crypts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "829b2cc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of unique cells in crypt data:\n",
      "Intestinal epithelium\n",
      "Performing differential expression analysis of Intestinal epithelium cells ⌛\n",
      "Subsetting 6 cells\n",
      "Removing ambient genes ⌛\n",
      "Removing genes expressed in less than 10% of cells ⌛\n",
      "Running Limma ⌛\n",
      "Loading edgeR package\n",
      "Loading limma package\n",
      "Assigning data ⌛\n",
      "Assigning condition\n",
      "Assigning batch\n",
      "Formatting data\n",
      "Releveling condition\n",
      "Creating DGEList\n",
      "Deleting data\n",
      "Creating EList\n",
      "Assigning DGEList\n",
      "Deleting DGEList\n",
      "Creating design matrix\n",
      "Running Limma lmFit ⌛\n",
      "Running Limma eBayes ⌛\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning message:\n",
      "\n",
      "R[write to console]: 44 very small variances detected, have been offset away from zero \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting results\n",
      "Data stored in variable x\n",
      "Calculating absolute lfc\n",
      "Group2: HFD in tot: True\n",
      "Group1: CD in tot: True\n",
      "Writing to diff_exp_CD_vs_HFD_Intestinal epithelium ⌛\n",
      "Done ✅\n"
     ]
    }
   ],
   "source": [
    "adata_crypts_pb = sc.read_h5ad('/home/glennrd/Documents/Research_Project/RNA-seq_Analysis/python/crypt/crypt_enriched_pseudobulk.h5ad')\n",
    "\n",
    "# Create batch annotation for limma (Crypts)\n",
    "adata_crypts_pb.obs['Batch'] = ['batch'+x.split('_')[1] for x in adata_crypts_pb.obs['Sample']]\n",
    "adata_crypts_pb.obs['Batch'] = adata_crypts_pb.obs['Batch'].astype('category')\n",
    "\n",
    "cells_crypts_pb = adata_crypts_pb.obs['major_cell_types'].unique()\n",
    "print('List of unique cells in crypt data:')\n",
    "for cell in cells_crypts_pb:\n",
    "    print(cell)\n",
    "\n",
    "# Define variables for adata_crypts_pb\n",
    "list_cells_crypts_pb = cells_crypts_pb\n",
    "dataset_crypts_pb = 'crypt_pb'\n",
    "group_crypts_pb = 'major_cell_types'\n",
    "\n",
    "# Run differential expression analysis for adata_villi\n",
    "differential_expression_analysis(list_cells_crypts_pb, adata_crypts_pb, dataset_crypts_pb, group_crypts_pb)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
